### PR TITLE
Add Login with Discord button

### PR DESCRIPTION
**Pull request by qsaml**
# SCE Authentication with Discord

Implements a login button to connect your SCE account to your Discord account.

# How to test
- in `api/config`, locate `config.example.json` and create a copy of that file, renaming it to `config.json`
- in this new file created, obtain `CLIENT_ID` and `CLIENT_SECRET` from sce-dev discord, then paste them in the respective fields (this is in `discordApiKeys`)
- make sure mongodb is running, then in the terminal, open the server and frontend (`npm run server` then `npm start`)
- login with sce account, navigate to the profile page, you'll see the page look like this
image
- click on "connect with discord" and follow the prompts. the authentication page will open in a new tab
- login with your discord account, then close the discord tab if it displays "Authorization successful! You may now close this window or tab"
- refresh the profile page. you'll see the page now looking like this 
image
- note! if you have a discord account linked already, the profile screen will appear as it does above. since a person's username (and discriminator, with discord nitro) can be changed at will, it is recommended that each time that person updates said information, they also update with sce their new username (and discriminator). discordID will not change.

# Final notes

- upon the return of the official sce website, instead of redirecting to discord authorization page, it will instead redirect back to the profile page.

### CODE REVIEW

File: api/config/config.example.json [hunk 1]
```diff
@@ -15,5 +15,11 @@
   "ledSignIp": "192.168.4.1",
   "GENERAL_API_URL": "http://localhost:8080/",
   "LOGGING_API_URL": "http://localhost:8081/",
-  "MAILER_API_URL": "http://localhost:8082/"
+  "MAILER_API_URL": "http://localhost:8082/",
+  "discordApiKeys": {
+    "CLIENT_SECRET": "XXXXXXXXXXXXXXXXXX",
+    "CLIENT_ID": "XXXXXXXXXXXXXXXXXX",
+    "REDIRECT_URI": "http%3A%2F%2Flocalhost%3A8080%2Fapi%2Fuser%2Fcallback&response_type=code&scope=identify",
```
Comment (evanugarte) [inline api/config/config.example.json:22]: keep in mind that eventually we will no longer redirect users to `localhost` but instead `sce.sjsu.edu`

File: api/main_endpoints/routes/User.js [hunk 2]
```diff
@@ -5,6 +5,9 @@ const router = express.Router();
 const passport = require('passport');
 require('../util/passport')(passport);
 const User = require('../models/User.js');
+const fetch = require('node-fetch');
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:8]: lets stick to `axios.get` and not use `fetch`

File: api/main_endpoints/routes/User.js [hunk 3]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
   });
 });
 
+// Verifies the users session if they have an active jwtToken.
+// Used on the inital load of root '/'
+// Returns the name and accesslevel of the user w/ the given access token
+router.post('/verify', function(req, res) {
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:213]: can you remove lines `210 - 225`? we already have this [here](https://github.com/SCE-Development/Core-v4/blob/e465da89e0719bd60f5c009267e017242d9ceaff/api/main_endpoints/routes/Auth.js#L151)

File: api/main_endpoints/routes/User.js [hunk 4]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+  } else {
+    res.sendStatus(UNAUTHORIZED);
+  }
+});
+
+router.get('/callback', async function(req, res) {
+  const code = req.query.code;
+  const email = req.query.state;
+  axios
+    .post('https://discord.com/api/oauth2/token',
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:230]: perhaps make a helper function in `main_endpoints/util` to genreate this link

File: api/main_endpoints/routes/User.js [hunk 5]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+        }
+      })
+        .then(loginResponse => {
+          User.findOne({email:email})
+            .then((user) => {
+              user.discordID = loginResponse.data.id;
+              user.save();
+              return res.redirect('https://discord.com/oauth2/authorized');
+            })
+            .catch(() => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:248]: you can keep the catch on line 252 and it will catch any errors within these brackets

File: api/main_endpoints/routes/User.js [hunk 6]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+      axios.get('https://discord.com/api/users/@me', {
+        headers: {
+          'Authorization': `Bearer ${response.data.access_token}`,
+        }
+      })
+        .then(loginResponse => {
+          User.findOne({email:email})
+            .then((user) => {
+              user.discordID = loginResponse.data.id;
+              user.save();
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:245]: is this asyncronous? you may want to `await` it to ensure the save worked

File: api/main_endpoints/routes/User.js [hunk 7]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+      `&grant_type=client_credentials&code=${code}` +
+      `&redirect_uri=${discordApiKeys.REDIRECT_URI}`)
+    .then(response => {
+      axios.get('https://discord.com/api/users/@me', {
+        headers: {
+          'Authorization': `Bearer ${response.data.access_token}`,
+        }
+      })
+        .then(loginResponse => {
+          User.findOne({email:email})
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:242]:
```suggestion
          User.findOne({ email })
```

File: api/main_endpoints/routes/User.js [hunk 8]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+  axios
+    .post('https://discord.com/api/oauth2/token',
+      `client_id=${discordApiKeys.CLIENT_ID}` +
+      `&client_secret=${discordApiKeys.CLIENT_SECRET}` +
+      `&grant_type=client_credentials&code=${code}` +
+      `&redirect_uri=${discordApiKeys.REDIRECT_URI}`)
+    .then(response => {
+      axios.get('https://discord.com/api/users/@me', {
+        headers: {
+          'Authorization': `Bearer ${response.data.access_token}`,
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:238]:
```suggestion
          Authorization: `Bearer ${response.data.access_token}`,
```

File: api/main_endpoints/routes/User.js [hunk 9]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+  if (decoded) {
+    res.status(OK).send(decoded);
+  } else {
+    res.sendStatus(UNAUTHORIZED);
+  }
+});
+
+router.get('/callback', async function(req, res) {
+  const code = req.query.code;
+  const email = req.query.state;
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:228]: for these lines you can just send over params like `code` and `email` instead of `code` and `state`, then u can do

```js
const { code, email } = req.query;
```

read about [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)

File: api/main_endpoints/routes/User.js [hunk 10]
```diff
@@ -199,6 +207,73 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+  const email = req.body.email;
+  if (!email) {
+    return res.sendStatus(BAD_REQUEST);
+  }
+  return res.send('https://discord.com/api/oauth2/authorize?client_id=' +
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:271]: use that helper function i mentioned before

File: src/APIFunctions/User.js [hunk 11]
```diff
@@ -34,6 +34,19 @@ export async function updateLastLoginDate(email, token) {
 }
 
 
+export async function connectToDiscord(email, token) {
+  let status = new UserApiResponse();
+  // eslint-disable-next-line
+  await axios.post('api/user/connectToDiscord', {email:email, token:token})
```
Comment (evanugarte) [inline src/APIFunctions/User.js:40]:
```suggestion
  await axios.post('api/user/connectToDiscord', { email, token })
```

File: src/APIFunctions/User.js [hunk 12]
```diff
@@ -34,6 +34,19 @@ export async function updateLastLoginDate(email, token) {
 }
 
 
+export async function connectToDiscord(email, token) {
+  let status = new UserApiResponse();
+  // eslint-disable-next-line
```
Comment (evanugarte) [inline src/APIFunctions/User.js:39]: why is this here?

File: src/APIFunctions/User.js [hunk 13]
```diff
@@ -34,6 +34,19 @@ export async function updateLastLoginDate(email, token) {
 }
 
 
+export async function connectToDiscord(email, token) {
```
Comment (evanugarte) [inline src/APIFunctions/User.js:37]: wouldnt you want to put this into [`src/APIFunctions/Auth.js`](https://github.com/SCE-Development/Core-v4/blob/dev/src/APIFunctions/Auth.js)?

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 14]
```diff
@@ -1,9 +1,11 @@
+/* eslint-disable */
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:1]: nice try
```suggestion
```

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 15]
```diff
@@ -16,7 +18,8 @@ export default function ProfileCard(props) {
     // hash pass
     const salt = bcrypt.genSaltSync(10);
     const hashedPassword =
-      password.trim() === '' ? user.password : bcrypt.hashSync(password, salt);
+        // eslint-disable-next-line
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:21]: if this is here due to line length u can do
```js
password.trim() === '' ?
  user.password :
  bcrypt.hashSync(password, salt);
```

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 16]
```diff
@@ -67,8 +75,16 @@ export default function ProfileCard(props) {
             }
           </h3>
         ))}
+        <Button
+          className='discord-button'
+          onClick={buttonClick}>
+          <div className='center-text'>
+            <img alt='disLogo' src='https://www.freeiconspng.com/uploads/discord-black-icon-1.png'></img>
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:82]:
```suggestion
            <img alt='discord logo' src='https://www.freeiconspng.com/uploads/discord-black-icon-1.png'></img>
```

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 17]
```diff
@@ -67,8 +75,16 @@ export default function ProfileCard(props) {
             }
           </h3>
         ))}
+        <Button
+          className='discord-button'
+          onClick={buttonClick}>
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:80]:
```suggestion
          onClick={handleDiscordAuth}>
```

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 18]
```diff
@@ -48,6 +51,11 @@ export default function ProfileCard(props) {
     return style;
   }
 
+  async function buttonClick() {
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:54]:
```suggestion
  async function handleDiscordAuth() {
```

File: src/Routing.js [hunk 19]
```diff
@@ -35,6 +35,7 @@ export default function Routing({ appProps }) {
     appProps.user &&
     appProps.user.accessLevel >= membershipState.OFFICER;
   const signedInRoutes = [
+
```
Comment (evanugarte) [inline src/Routing.js:38]: revert the changes in this file if u dont need it

File: api/config/config.example.json (hunk 1 above)
Comment (evanugarte) [reply api/config/config.example.json:22]: you can keep it as is for now, but we should have checks for when were running in production or not

File: api/main_endpoints/routes/User.js (hunk 3 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:213]: I believe that was already there

File: api/main_endpoints/routes/User.js (hunk 5 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:248]: so the one on 248 isn't needed then?

File: api/main_endpoints/routes/User.js (hunk 4 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:230]: isn't link auto-generated by discord developer app I created? not sure how I'd implement this

File: api/main_endpoints/routes/User.js (hunk 6 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:245]: not sure the meaning of asyncronous, assuming that it's inside an `async` function? if not, `await` goes where?

File: src/APIFunctions/User.js (hunk 12 above)
Comment (qsaml) [reply src/APIFunctions/User.js:39]: was because of change previously mentioned

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 16 above)
Comment (qsaml) [reply src/Pages/Profile/MemberView/InfoCard.js:82]: how could I make it so that the length of the line doesn't trigger lint? putting `alt` and `src` on different lines doesn't help as `src` part still exceeds char limit

File: api/main_endpoints/routes/User.js (hunk 3 above)
Comment (evanugarte) [reply api/main_endpoints/routes/User.js:213]: then why is it green in this pr

File: api/main_endpoints/routes/User.js (hunk 4 above)
Comment (evanugarte) [reply api/main_endpoints/routes/User.js:230]: sorry this comment is for `api/main_endpoints/routes/User.js` line 231, not sure why it's here

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 16 above)
Comment (evanugarte) [reply src/Pages/Profile/MemberView/InfoCard.js:82]:
```jsx
<img
  alt='disLogo'
  src=`${DISCORD_ICON_URL}`
/>
```

File: api/main_endpoints/routes/User.js (hunk 4 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:230]: still not really understanding, I get you're talking about this part...
``client_id=${discordApiKeys.CLIENT_ID}` +
      `&client_secret=${discordApiKeys.CLIENT_SECRET}` +
      `&grant_type=client_credentials&code=${code}` +
      `&redirect_uri=${discordApiKeys.REDIRECT_URI}``
I don't get how you'd make a helper function because again, Discord is automatically generating this link

File: api/config/config.example.json (hunk 1 above)
Comment (qsaml) [reply api/config/config.example.json:22]: will change this when the site does indeed redirect to sce.sjsu.edu, have to do it in Discord as well

File: src/APIFunctions/User.js (hunk 13 above)
Comment (qsaml) [reply src/APIFunctions/User.js:37]: will switch it over and see if if breaks anything

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 16 above)
Comment (qsaml) [reply src/Pages/Profile/MemberView/InfoCard.js:82]: just had an oopsie here with spacing

File: api/main_endpoints/routes/User.js (hunk 9 above)
Comment (qsaml) [reply api/main_endpoints/routes/User.js:211]: tried and doesn't work that way, email and state are differently named

**Comment by hsorathia** [comment]
Not sure why this didn't work for me. My steps:

1. Start and run the server w/ mongo
  ```
  npm start
  npm run server
  mongo
  ```
2. Login to an account I made a while back.
1. Navigate to profile and click on the discord button.

Error messages:
image

image

**Comment by qsaml** [comment]
I will update with instructions as to how to get the api keys on my next push

**Comment by qsaml** [comment]
Future reviewers: make sure you're getting my code from `SamL/discordButton`

**Comment by jerrylee17** [comment]
Link this to an issue

File: api/main_endpoints/models/User.js [hunk 20]
```diff
@@ -36,6 +36,7 @@ const UserSchema = new Schema(
     },
     discordID: {
       type: String,
+      default: ''
     },
 
     // Users declared Major at SJSU
```
Comment (seemavora) [inline api/main_endpoints/models/User.js:42]: dont think u need this comment, remove

File: test/User.js [hunk 21]
```diff
@@ -14,11 +14,14 @@ const {
...
 const SceApiTester = require('./util/tools/SceApiTester');
+const discordModule = require('../api/main_endpoints/util/discord-connection');
 
 
 let app = null;
 let test = null;
+let sandbox = sinon.createSandbox();
 
 const expect = chai.expect;
 // tools for testing
```
Comment (seemavora) [inline test/User.js:27]: think u can remove this too

Comment (seemavora) [reply test/User.js:27]: also squash ur commits!

**Comment by qsaml** [comment]
Commits squashed and a few unneeded comments removed. Pretty sure these comments were already there as I didn't add them

**Comment by ThaiQ** [comment]
Rebase SAMMMM

File: api/main_endpoints/routes/Auth.js [hunk 22]
```diff
@@ -98,7 +98,7 @@ router.post('/login', function(req, res) {
                 // query
                 { email: user.email },
                 // update this field
-                { pagesPrinted: 0 }
+                { pagesPrinted: 0 },
```
Comment (evanugarte) [inline api/main_endpoints/routes/Auth.js:101]: not sure why this file has changes, id revert it

File: api/main_endpoints/routes/User.js [hunk 23]
```diff
@@ -199,6 +206,33 @@ router.post('/getPagesPrintedCount', (req, res) => {
...
+  return res.status(NOT_FOUND).send('Authorization unsuccessful!');
+});
+
+router.post('/connectToDiscord', function(req, res) {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+  const email = req.body.email;
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js:225]: i would put this above `checkIfTokenSent` on line 220. seems like theres no point of checking token stuff if theres no email to begin with

File: api/main_endpoints/util/discord-connection.js [hunk 24]
```diff
@@ -0,0 +1,39 @@
+const User = require('../models/User');
+const axios = require('axios');
+const {
+  discordApiKeys
+} = require('../../config/config.json');
+
+function loginWithDiscord(code, email) {
+  return axios
+    .post('https://discord.com/api/oauth2/token',
```
Comment (evanugarte) [inline api/main_endpoints/util/discord-connection.js:9]: `axios.post` supports key value pairs. you can do
```js
.post('https://discord.com/api/oauth2/token',
  client_id: discordApiKeys.CLIENT_ID,
  client_secret: discordApiKeys.CLIENT_SECRET,
  grant_type: 'authorization_code',
  code,
  redirect_uri: discordApiKeys.REDIRECT_URI
})
```

File: api/main_endpoints/util/discord-connection.js [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+              user.save();
+              return true;
+            });
+        })
+        .catch(() => {
+          return false;
+        });
+    })
+    .catch(() => {
+      return false;
```
Comment (evanugarte) [inline api/main_endpoints/util/discord-connection.js:33]: are you trying to return `true`/`false` back to wherever `loginWithDiscord` was called? If so, this won't behave as you expected.

You should return a `Promise` for this instead. `resolve(true)` where u originally returned it and `reject(false)` where it was originally a `return`

See [`api/google_api/util/SceGoogleApiHandler.js:52`](https://github.com/SCE-Development/Core-v4/blob/55ad626d991b3048a917b63abb68620a89911dad/api/google_api/util/SceGoogleApiHandler.js#L52) for an example.

File: src/Pages/Profile/MemberView/InfoCard.js [hunk 26]
```diff
@@ -4,19 +4,23 @@ import './profile-modifier.css';
...
 import { editUser } from '../../../APIFunctions/User';
+import { connectToDiscord } from '../../../APIFunctions/User';
 const pic = require('./getPicBySeason');
 const bcrypt = require('bcryptjs');
 
 export default function ProfileCard(props) {
   const [password, setPassword] = useState('New Password');
   const [confirmPass, setConfirmPass] = useState('Confirming New Password');
   const [user, setUser] = useState('');
+  const iLink = 'https://www.freeiconspng.com/uploads/discord-black-icon-1.png';
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/InfoCard.js:15]:
```suggestion
  const discordIcon = 'https://www.freeiconspng.com/uploads/discord-black-icon-1.png';
```

File: src/Routing.js (hunk 19 above)
Comment (evanugarte) [reply src/Routing.js:38]: ^

File: test/User.js [hunk 27]
```diff
@@ -270,4 +273,65 @@ describe('User', () => {
...
+        token: 'Invalid token'
+      };
+      const result = await test.sendPostRequest(
+        '/api/user/connectToDiscord', user);
+      expect(result).to.have.status(UNAUTHORIZED);
+    });
+    it('Should return statusCode 400 if an incorrect or no ' +
+      'email was used', async () => {
+      const user = {
+        token: token
```
Comment (evanugarte) [inline test/User.js:299]:
```suggestion
        token
```

File: test/User.js [hunk 28]
```diff
@@ -270,4 +273,65 @@ describe('User', () => {
...
+      setTokenStatus(true);
+      const result = await test.sendPostRequestWithToken(
+        token, '/api/user/connectToDiscord', user);
+      expect(result).to.have.status(BAD_REQUEST);
+    });
+    it('Should return statusCode 200 ' +
+      'if Discord connection was successful', async () => {
+      const user = {
+        email: 'a@b.c',
+        token: token
```
Comment (evanugarte) [inline test/User.js:310]:
```suggestion
        token
```

File: api/main_endpoints/routes/Auth.js (hunk 22 above)
Comment (qsaml) [reply api/main_endpoints/routes/Auth.js:101]: here I had to manually make the edit to fix one of the rebases because I had a panic

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 26 above)
Comment (qsaml) [reply src/Pages/Profile/MemberView/InfoCard.js:15]: breaks lint if it happens (80 chars)

File: test/User.js (hunk 27 above)
Comment (qsaml) [reply test/User.js:299]: what about for the rest of the tests that are written in this way? those necessary to change as well?

Comment (evanugarte) [reply test/User.js:299]: changing just these is fine

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 26 above)
Comment (evanugarte) [reply src/Pages/Profile/MemberView/InfoCard.js:15]:
```js
const iconName =
  'link/to/file.jpg'
```

File: api/main_endpoints/routes/Auth.js (hunk 22 above)
Comment (evanugarte) [reply api/main_endpoints/routes/Auth.js:101]: ok just change it back

File: src/Pages/Profile/MemberView/InfoCard.js (hunk 26 above)
Comment (qsaml) [reply src/Pages/Profile/MemberView/InfoCard.js:15]: didn't catch that way, alright

File: api/main_endpoints/util/discord-connection.js (hunk 24 above)
Comment (qsaml) [reply api/main_endpoints/util/discord-connection.js:9]: errors near the colons as shown below
image

Comment (evanugarte) [reply api/main_endpoints/util/discord-connection.js:9]: add an opening bracket on line 9. ask any more questions in discord please

Comment (qsaml) [reply api/main_endpoints/util/discord-connection.js]: edit: tried this approach and the code will not work at all when written this way. has to be written as it was before

**Comment by qsaml** [comment]
> Can we show something different it the user successfully signs in? Do sum like
> 
> ```
> Discord verified! ✔️
> Username: <username>
> ```
> 
> The `Connect with Discord` button could change to `Change Discord Account` or something in case the user switches accounts.

this approach works, but recall that since the discord link opens in a new tab, the change won't immediately reflect unless user refreshes the profile page.

I can make this redirect back to the profile page once the actual website is running

File: api/config/config.example.json (hunk 1 above)
Comment (evanugarte) [inline api/config/config.example.json:22]: consider making 2 fields like

```suggestion
    "REDIRECT_URI_DEV": "http%3A%2F%2Flocalhost%3A8080%2Fapi%2Fuser%2Fcallback&response_type=code&scope=identify",
    "REDIRECT_URI_PROD": "http%3A%2F%2Fsce.sjsu.edu%2Fapi%2Fuser%2Fcallback&response_type=code&scope=identify",
```

then you can conditionally redirect to either dev or prod if `process.env.NODE_ENV === 'production'`

File: src/Pages/Profile/MemberView/Profile.jsx [hunk 29]
```diff
@@ -24,6 +24,7 @@ export default class Profile extends Component {
   async componentDidMount() {
     const response = await searchUserByEmail(this.props.user.email,
       this.props.user.token)
+      console.log(response);
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/Profile.jsx:27]:
```suggestion
```

File: api/main_endpoints/util/discord-connection.js [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+const axios = require('axios');
+const {
+  discordApiKeys
+} = require('../../config/config.json');
+
+async function loginWithDiscord(code, email) {
+  return new Promise((resolve, reject) => {
+    axios
+      .post('https://discord.com/api/oauth2/token',
+        `&client_id=${discordApiKeys.CLIENT_ID}` +
```
[the remark this answers is not in the document]
Comment (qsaml) [reply api/main_endpoints/util/discord-connection.js:11]: the code broke when implementing it that way

File: api/config/config.example.json [hunk 31]
```diff
@@ -15,5 +15,11 @@
...
+    "REFRESH_TOKEN": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
```
[the remark this answers is not in the document]
Comment (qsaml) [reply api/config/config.example.json:23]: if we work with linking the discord bot to core-v4 this may be necessary

File: api/config/config.example.json (hunk 1 above)
Comment (qsaml) [reply api/config/config.example.json:22]: alright, will update the discord dev-app info with the prod link when the website comes back. side note - where would that line of code `process.env.NODE_ENV === 'production'` go?

File: api/main_endpoints/util/discord-connection.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (evanugarte) [reply api/main_endpoints/util/discord-connection.js:11]: remember you can ignore sections of code for style...
`/* eslint-diable camelcase */

**Comment by qsaml** [comment]
> lets not upload src/Pages/Profile/MemberView/Image/discord.jpg. please link it via URL

going to delete that file

File: src/Pages/Profile/MemberView/Profile.jsx (hunk 29 above)
Comment (qsaml) [reply src/Pages/Profile/MemberView/Profile.jsx:27]: didn't catch that

File: api/config/config.example.json (hunk 1 above)
Comment (evanugarte) [reply api/config/config.example.json:22]: if you change the axios.post back to use [this](https://github.com/SCE-Development/Core-v4/pull/562#discussion_r464208028)

you can do

```js
const runningInProduciton = process.env.NODE_ENV === 'production';
// stuff ...
// in axios.post
{
  redirect_uri: runningInProduciton ? REDIRECT_URI_PROD  : REDIRECT_URI_DEV
}
```

File: api/main_endpoints/util/discord-connection.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (qsaml) [reply api/main_endpoints/util/discord-connection.js:11]: but it wasn't a lint error, stuff just didn't store properly when I tried adding the new fields

[the remark this answers is not in the document]
Comment (qsaml) [reply api/main_endpoints/util/discord-connection.js:11]: more specifically, this popped up in browser console
image
